File: backend/src/models/Product.model.js
const mongoose = require('mongoose');

const productSchema = new mongoose.Schema(
  {
    shopId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Shop',
      required: true,
    },
    name: {
      type: String,
      required: [true, 'Product name is required'],
      trim: true,
      maxlength: [200, 'Product name cannot exceed 200 characters'],
    },
    localName: {
      type: String,
      trim: true,
      maxlength: [200, 'Local name cannot exceed 200 characters'],
    },
    description: {
      type: String,
      maxlength: [1000, 'Description cannot exceed 1000 characters'],
    },
    category: {
      type: String,
      trim: true,
      maxlength: [100, 'Category cannot exceed 100 characters'],
    },
    subCategory: {
      type: String,
      trim: true,
    },
    sku: {
      type: String,
      trim: true,
    },
    barcode: {
      type: String,
      trim: true,
    },
    unit: {
      type: String,
      enum: ['piece', 'kg', 'gram', 'liter', 'ml', 'meter', 'dozen', 'packet', 'box', 'bundle', 'other'],
      default: 'piece',
    },
    purchasePrice: {
      type: Number,
      required: [true, 'Purchase price is required'],
      min: [0, 'Purchase price cannot be negative'],
    },
    sellingPrice: {
      type: Number,
      required: [true, 'Selling price is required'],
      min: [0, 'Selling price cannot be negative'],
    },
    mrp: {
      type: Number,
      min: [0, 'MRP cannot be negative'],
    },
    taxRate: {
      type: Number,
      default: 0,
      min: [0, 'Tax rate cannot be negative'],
      max: [100, 'Tax rate cannot exceed 100%'],
    },
    currentStock: {
      type: Number,
      default: 0,
      min: [0, 'Stock cannot be negative'],
    },
    minStockLevel: {
      type: Number,
      default: 5,
      min: [0, 'Minimum stock level cannot be negative'],
    },
    maxStockLevel: {
      type: Number,
      default: 1000,
    },
    reorderPoint: {
      type: Number,
      default: 10,
    },
    images: [
      {
        url: String,
        isPrimary: {
          type: Boolean,
          default: false,
        },
        uploadedAt: {
          type: Date,
          default: Date.now,
        },
      },
    ],
    supplier: {
      name: String,
      phone: String,
      address: String,
    },
    expiryDate: {
      type: Date,
      default: null,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    tags: [String],
    lastRestockedAt: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

// Indexes
productSchema.index({ shopId: 1, name: 'text', localName: 'text' });
productSchema.index({ shopId: 1, category: 1 });
productSchema.index({ shopId: 1, isActive: 1 });
productSchema.index({ shopId: 1, barcode: 1 }, { sparse: true });
productSchema.index({ shopId: 1, sku: 1 }, { sparse: true });
productSchema.index({ shopId: 1, currentStock: 1, minStockLevel: 1 });

// Virtual for profit margin
productSchema.virtual('profitMargin').get(function () {
  if (this.purchasePrice === 0) return 100;
  return (((this.sellingPrice - this.purchasePrice) / this.purchasePrice) * 100).toFixed(2);
});

// Virtual for low stock status
productSchema.virtual('isLowStock').get(function () {
  return this.currentStock <= this.minStockLevel;
});

// Virtual for out of stock status
productSchema.virtual('isOutOfStock').get(function () {
  return this.currentStock === 0;
});

// Virtual for stock value
productSchema.virtual('stockValue').get(function () {
  return this.currentStock * this.purchasePrice;
});

productSchema.set('toJSON', { virtuals: true });
productSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model('Product', productSchema);

File: khata_setu_app/lib/core/data/models/product_model.dart
import 'package:hive/hive.dart';

part 'product_model.g.dart';

@HiveType(typeId: 5)
class ProductModel extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  String name;

  @HiveField(2)
  String? localName;

  @HiveField(3)
  String? description;

  @HiveField(4)
  String category;

  @HiveField(5)
  String? subCategory;

  @HiveField(6)
  String? sku;

  @HiveField(7)
  String? barcode;

  @HiveField(8)
  String unit; // piece, kg, gram, liter, ml, meter, dozen, packet, box, bundle, other

  @HiveField(9)
  double purchasePrice;

  @HiveField(10)
  double sellingPrice;

  @HiveField(11)
  double? mrp;

  @HiveField(12)
  double taxRate;

  @HiveField(13)
  double currentStock;

  @HiveField(14)
  double minStockLevel;

  @HiveField(15)
  double maxStockLevel;

  @HiveField(16)
  double reorderPoint;

  @HiveField(17)
  String? image;

  @HiveField(18)
  bool isActive;

  @HiveField(19)
  DateTime createdAt;

  @HiveField(20)
  DateTime? lastRestockedAt;

  @HiveField(21)
  bool synced;

  @HiveField(22)
  List<String> tags;

  @HiveField(23)
  String? supplierName;

  @HiveField(24)
  String? supplierPhone;

  @HiveField(25)
  DateTime? expiryDate;

  ProductModel({
    required this.id,
    required this.name,
    this.localName,
    this.description,
    this.category = 'General',
    this.subCategory,
    this.sku,
    this.barcode,
    this.unit = 'piece',
    required this.purchasePrice,
    required this.sellingPrice,
    this.mrp,
    this.taxRate = 0,
    this.currentStock = 0,
    this.minStockLevel = 5,
    this.maxStockLevel = 1000,
    this.reorderPoint = 10,
    this.image,
    this.isActive = true,
    DateTime? createdAt,
    this.lastRestockedAt,
    this.synced = false,
    List<String>? tags,
    this.supplierName,
    this.supplierPhone,
    this.expiryDate,
  })  : createdAt = createdAt ?? DateTime.now(),
        tags = tags ?? [];

  // ─── Computed ────────────────────────────────────────────────

  bool get isLowStock => currentStock > 0 && currentStock <= minStockLevel;
  bool get isOutOfStock => currentStock <= 0;
  double get profitMargin =>
      sellingPrice > 0 ? ((sellingPrice - purchasePrice) / sellingPrice) * 100 : 0;
  double get stockValue => currentStock * purchasePrice;

  /// Emoji based on category for display
  String get categoryEmoji {
    switch (category.toLowerCase()) {
      case 'grocery':
      case 'kirana':
        return '🛒';
      case 'dairy':
        return '🥛';
      case 'snacks':
        return '🍪';
      case 'beverages':
        return '🥤';
      case 'personal care':
        return '🧴';
      case 'stationery':
        return '✏️';
      case 'household':
        return '🏠';
      case 'medical':
      case 'medicine':
        return '💊';
      case 'electronics':
        return '📱';
      case 'clothing':
        return '👕';
      case 'hardware':
        return '🔧';
      case 'fruits':
      case 'vegetables':
        return '🥬';
      default:
        return '📦';
    }
  }

  ProductModel copyWith({
    String? id,
    String? name,
    String? localName,
    String? description,
    String? category,
    String? subCategory,
    String? sku,
    String? barcode,
    String? unit,
    double? purchasePrice,
    double? sellingPrice,
    double? mrp,
    double? taxRate,
    double? currentStock,
    double? minStockLevel,
    double? maxStockLevel,
    double? reorderPoint,
    String? image,
    bool? isActive,
    DateTime? createdAt,
    DateTime? lastRestockedAt,
    bool? synced,
    List<String>? tags,
    String? supplierName,
    String? supplierPhone,
    DateTime? expiryDate,
  }) {
    return ProductModel(
      id: id ?? this.id,
      name: name ?? this.name,
      localName: localName ?? this.localName,
      description: description ?? this.description,
      category: category ?? this.category,
      subCategory: subCategory ?? this.subCategory,
      sku: sku ?? this.sku,
      barcode: barcode ?? this.barcode,
      unit: unit ?? this.unit,
      purchasePrice: purchasePrice ?? this.purchasePrice,
      sellingPrice: sellingPrice ?? this.sellingPrice,
      mrp: mrp ?? this.mrp,
      taxRate: taxRate ?? this.taxRate,
      currentStock: currentStock ?? this.currentStock,
      minStockLevel: minStockLevel ?? this.minStockLevel,
      maxStockLevel: maxStockLevel ?? this.maxStockLevel,
      reorderPoint: reorderPoint ?? this.reorderPoint,
      image: image ?? this.image,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      lastRestockedAt: lastRestockedAt ?? this.lastRestockedAt,
      synced: synced ?? this.synced,
      tags: tags ?? this.tags,
      supplierName: supplierName ?? this.supplierName,
      supplierPhone: supplierPhone ?? this.supplierPhone,
      expiryDate: expiryDate ?? this.expiryDate,
    );
  }
}
